providers/implementations/kem/template_kem.c

```diff
@@ -0,0 +1,183 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/err.h>
+#include <openssl/proverr.h>
+#include "prov/provider_ctx.h"
+#include "prov/implementations.h"
+#include "prov/securitycheck.h"
+#include "prov/providercommon.h"
+
+extern const OSSL_DISPATCH ossl_template_asym_kem_functions[];
+
+#if defined(NDEBUG) || defined(OPENSSL_NO_STDIO)
+static void debug_print(char *fmt, ...)
+{
+}
+
+#else
+static void debug_print(char *fmt, ...)
+{
+    char out[1000];
+    va_list argptr;
+
+    va_start(argptr, fmt);
+    vsprintf(out, fmt, argptr);
+    va_end(argptr);
+    if (getenv("TEMPLATEKEM"))
+        fprintf(stderr, "TEMPLATE_KEM: %s", out);
+}
+#endif
+
+typedef struct {
+    OSSL_LIB_CTX *libctx;
+    /* some key struct */
+    int op;
+} PROV_TEMPLATE_CTX;
+
+static OSSL_FUNC_kem_newctx_fn template_newctx;
+static OSSL_FUNC_kem_encapsulate_init_fn template_encapsulate_init;
+static OSSL_FUNC_kem_encapsulate_fn template_encapsulate;
+static OSSL_FUNC_kem_decapsulate_init_fn template_decapsulate_init;
+static OSSL_FUNC_kem_decapsulate_fn template_decapsulate;
+static OSSL_FUNC_kem_freectx_fn template_freectx;
+static OSSL_FUNC_kem_set_ctx_params_fn template_set_ctx_params;
+
+static void *template_newctx(void *provctx)
+{
+    PROV_TEMPLATE_CTX *ctx = OPENSSL_zalloc(sizeof(*ctx));
+
+    debug_print("newctx called\n");
+    if (ctx == NULL)
+        return NULL;
+    ctx->libctx = PROV_LIBCTX_OF(provctx);
+
+    debug_print("newctx returns %p\n", ctx);
+    return ctx;
+}
+
+static void template_freectx(void *vctx)
+{
+    PROV_TEMPLATE_CTX *ctx = (PROV_TEMPLATE_CTX *)vctx;
+
+    debug_print("freectx %p\n", ctx);
+    OPENSSL_free(ctx);
+}
+
+static int template_init(void *vctx, int operation, void *vkey, void *vauth,
+                         ossl_unused const OSSL_PARAM params[])
+{
+    PROV_TEMPLATE_CTX *ctx = (PROV_TEMPLATE_CTX *)vctx;
+
+    debug_print("init %p / %p\n", ctx, vkey);
+    if (!ossl_prov_is_running())
+        return 0;
+
+    /* check and fill in reference to key */
+    ctx->op = operation;
+    debug_print("init OK\n");
+    return 1;
+}
+
+static int template_encapsulate_init(void *vctx, void *vkey,
+                                     const OSSL_PARAM params[])
+{
+    return template_init(vctx, EVP_PKEY_OP_ENCAPSULATE, vkey, NULL, params);
+}
+
+static int template_decapsulate_init(void *vctx, void *vkey,
+                                     const OSSL_PARAM params[])
+{
+    return template_init(vctx, EVP_PKEY_OP_DECAPSULATE, vkey, NULL, params);
+}
+
+static int template_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    PROV_TEMPLATE_CTX *ctx = (PROV_TEMPLATE_CTX *)vctx;
+
+    debug_print("set ctx params %p\n", ctx);
+    if (ctx == NULL)
+        return 0;
+    if (params == NULL)
```

Comment:
```suggestion
    if (ossl_param_is_empty(params))
 ```
 Comes from..
 #include "prov/providercommon.h"

Reply:
Hmpf -- this PR predates the merge of https://github.com/openssl/openssl/commit/f5981c9629667a5a5d63cf1f88903ee6b54a45e3... Let me re-base and merge....

Reply:
new API now used.